system-addon/test/functional/mochitest/browser_topsites_section.js

```diff
@@ -0,0 +1,64 @@
+"use strict";
+
+registerCleanupFunction(async () => {
+  Services.prefs.clearUserPref("browser.newtabpage.activity-stream.default.sites");
+  Services.prefs.clearUserPref("browser.newtabpage.pinned");
+  await clearHistoryAndBookmarks();
+});
+
+// Check TopSites edit modal and overlay show up.
+test_newtab({
+  async before() {
+    await clearHistoryAndBookmarks();
+  },
+  // it should be able to click the topsites edit button to reveal the edit topsites modal and overlay.
+  test: function topsites_edit() {
+    const topsitesEditBtn = content.document.querySelector(".edit-topsites-button button");
+    topsitesEditBtn.click();
+
+    let found = content.document.querySelector(".edit-topsites");
+    ok(found && !found.hidden, "Should find a visible topsites edit menu");
+
+    found = content.document.querySelector(".modal-overlay");
+    ok(found && !found.hidden, "Should find a visible overlay");
+  }
+});
+
+// Test pin/unpin context menu options.
+test_newtab({
+  async before() {
+    // The pref for TopSites is empty by default.
+    Services.prefs.setStringPref("browser.newtabpage.activity-stream.default.sites", "https://www.youtube.com/,https://www.facebook.com/,https://www.amazon.com/,https://www.reddit.com/,https://www.wikipedia.org/,https://twitter.com/");
+    await clearHistoryAndBookmarks();
+  },
+  // it should pin the website when we click the first option of the topsite context menu.
+  test: async function topsites_pin_unpin() {
+    // There are only topsites on the page, the selector with find the first topsite menu button.
+    const topsiteContextBtn = content.document.querySelector(".context-menu-button");
+    topsiteContextBtn.click();
+
+    const contextMenu = content.document.querySelector(".context-menu");
+    ok(contextMenu && !contextMenu.hidden, "Should find a visible topsite context menu");
+
+    const pinUnpinTopsiteBtn = contextMenu.querySelector(".context-menu-item a");
+    // Pin the topsite.
+    pinUnpinTopsiteBtn.click();
+
+    // Need to wait for pin action.
+    await ContentTaskUtils.waitForCondition(() => content.document.querySelector(".icon-pin-small"),
+      "No highlights cards found.");
```

Comment:
Wrong failure message.

This makes me wonder if we want some helper that gets injected into each of our content tests, so we could just do `await waitFor(".icon-pin-small");` or `await waitForNot(…)` Looks like `ContentTaskUtils` is directly imported: https://searchfox.org/mozilla-central/source/testing/mochitest/BrowserTestUtils/content/content-task.js#10

We don't need to fix this refactoring here. I'll try asking around about best practices in adding helper functions to the content task.